portfolio/src/main/java/com/google/sps/servlets/MarkerServlet.java

```diff
@@ -99,4 +125,79 @@ public void storeMarker(Marker marker) {
 
     datastore.put(markerEntity);
   }
+
+  public boolean repeatMarkers(double reportLat, double reportLng, String reportDate,
+   String reportTime, String reportCrime) {
+    Query query = new Query(ENTITY_TITLE);
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+      double lat = (double) entity.getProperty("lat");
+      double lng = (double) entity.getProperty("lng");
+      String crime = (String) entity.getProperty("crimeType");
+      String date = (String) entity.getProperty("date");
+      String time = (String) entity.getProperty("time");
+
+      if (sameLocation(lat, lng, reportLat, reportLng)) 
+        if (sameDate(reportDate, date)) 
+          if (sameTime(reportTime, time)) 
+            if (sameCrime(reportCrime, crime))
+              return true;
+    }
+    return false;
+  }
+
+  public boolean sameLocation(double markerLat, double markerLng, double reportLat, double reportLng) {
+    double R = 6371.0710; // Radius of the Earth in kilometers
+    double rlat1 = markerLat * (Math.PI/180); // Convert degrees to radians
+    double rlat2 = reportLat * (Math.PI/180); // Convert degrees to radians
+    double difflat = rlat2-rlat1; // Radian difference (latitudes)
+    double difflon = (reportLng-markerLng) * (Math.PI/180); // Radian difference (longitudes)
+
+    double d = 2 * R * Math.asin(Math.sqrt(Math.sin(difflat/2)*Math.sin(difflat/2)+Math.cos(rlat1)*Math.cos(rlat2)*Math.sin(difflon/2)*Math.sin(difflon/2)));
+    /* If distance is less than 10 meters */
+    return d < 0.01;
+  }
+
+  public boolean sameDate(String date1, String date2) {
+    return date1.equals(date2);
+  }
+
+  public boolean sameTime(String time1, String time2) {
+    if (time1 == null || time2 == null){
```

Comment:
Do we ever expect to see NULL value?
It seems to me any NULL value we probably should fail with bad request.

Reply:
The user can submit a report without having to enter all possible fields, like time and date, but now I'm not sure if it would be null or a string with no characters